RBD Plugin: Implement Attacher/Detacher interfaces.

cofyc opened the pull request:
**What this PR does / why we need it**:

This PR continues @rootfs 's work in #33660. It implements volume.Attacher/Volume.Detacher interfaces to resolve RBD image locking and makes RBD plugin more robust.

Summary of interfaces and what they do for RBD plugin:

- Attacher.Attach(): does nothing
- Attacher.VolumesAreAttached(): method to query volume attach status
- Attacher.GetDeviceMountPath(): method to get device mount path 
- ~~Attacher.WaitForAttach(): kubelet maps the image on the node (and lock the image if needed)~~
- Attacher.WaitForAttach(): kubelet maps the image on the node (for backward compatibility, check if the image is still used by nodes running old kubelet)
- Attacher.MountDevice(): kubelet mounts device at the device mount path
- ~~Detacher.UnmountDevice: kubelet unmounts device from the device mount path (currently, we need to unmaps image from the node here) (and unlock the image if needed)~~
- Detacher.UnmountDevice: kubelet unmounts device from the device mount path and unmaps image from the node (for backward compatibility, clean old rbd.json file and previous lock acquired by old kubelets)
- Detacher.Detach(): does nothing

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

fixes #50142.

**Special notes for your reviewer**:

RBD changes:

  1) Modify rbdPlugin to implement volume.AttachableVolumePlugin interface.
  2) Add rbdAttacher/rbdDetacher structs to implement
  volume.Attacher/Detacher interfaces.
  3) Add mount.SafeFormatAndMount/mount.Exec fields to rbdPlugin, and setup them in
  rbdPlugin.Init for later uses. Attacher/Mounter/Unmounter/Detacher
  reference rbdPlugin to use mounter and exec. This simplifies
  code.
  4) Add testcase struct to abstract RBD Plugin test case, etc.
  5) Add newRBD constructor to unify rbd struct initialization.

Non-RBD changes:

  1) Fix FakeMounter.IsLikelyNotMountPoint to return ErrNotExist if the
  directory does not exist. Mounter.IsLikelyNotMountPoint interface
  requires this, and RBD plugin depends on it.
  2) ~~Extend Detacher.Detach method to pass `*volume.Spec`, RBD plugin
  needs it to detach device from the node.~~
  3) ~~Extend Volume.Spec struct to include namespace string, RBD Plugin needs
  it to locate objects (e.g. secrets) in Pod's namespace.~~
  4) ~~Update RABC bootstrap policy to allow
  `system:controller:attachdetach-controller` cluster role to get
  Secrets object. RBD attach/detach needs to access secrets object in
  Pod's namespace.~~

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```
NONE
```

rootfs commented:
@cofyc thanks! can you make into 2 commits, one for `Detach()` refactoring (and ping @saad-ali @jingxu97 @msau42) and one for rbd attach/detach interface?

rootfs commented:
/ok-to-test

Commit:
RBD Plugin: Implement Attacher/Detacher interfaces.

1) Modify rbdPlugin to implement volume.AttachableVolumePlugin
   interface.
2) Add rbdAttacher/rbdDetacher structs to implement
   volume.Attacher/Detacher interfaces.
3) Add mount.SafeFormatAndMount/mount.Exec fields to rbdPlugin, and
   setup them in rbdPlugin.Init for later uses.
   Attacher/Mounter/Unmounter/Detacher reference rbdPlugin to use mounter
   and exec. This simplifies code.
4) Add testcase struct to abstract RBD Plugin test case, etc.
5) Add newRBD constructor to unify rbd struct initialization.

cofyc commented:
@rootfs

PR has been split into two commits:

- https://github.com/kubernetes/kubernetes/pull/51608/commits/d5f0bc7fa7c1d7d4c5c8af685e7c69a1eeff2010 contains Non-RBD changes as I described in [main text](https://github.com/kubernetes/kubernetes/pull/51608#issue-253914816). (ping @saad-ali @jingxu97 @msau42)
- https://github.com/kubernetes/kubernetes/pull/51608/commits/a26454813e3f04c057b99ef945276601e435c155 contains RBD changes.

cofyc commented:
/test pull-kubernetes-unit

cofyc commented:
/test pull-kubernetes-e2e-kops-aws

rootfs commented on pkg/volume/rbd/rbd.go:
```diff
@@ -409,15 +443,187 @@ type rbd struct {
+	if err != nil {
+		return "", err
+	}
+	err = attacher.manager.Fencing(*mounter, string(nodeName))
```
cannot fence here: containerized controller-master doesn't have rbd cli in it. Check gcr.io/google_containers/kube-controller-manager-amd64

rootfs commented:
@cofyc some directional changes: instead of fencing rbd map at controller master, please use existing use [isMultiAttachForbidden](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L139) to enfore RWO. But to keep backward compatibility, check rbd lock at node in WaitForAttach.

See https://github.com/kubernetes/kubernetes/pull/50334 and https://github.com/kubernetes/kubernetes/pull/48940 for reference.

Commit:
RBD Plugin: Prepare to implement Attacher/Detacher interfaces.

1) Fix FakeMounter.IsLikelyNotMountPoint to return ErrNotExist if the
   directory does not exist. Mounter.IsLikelyNotMountPoint interface
   requires this, and RBD plugin depends on it.

cofyc commented:
@rootfs 
Thanks!
This simplifies code very much. Only [1 non-rbd change](https://github.com/kubernetes/kubernetes/pull/51608/commits/81be7aefb778b1617f9cb9cf47eb918aa3ecca30) is needed (actually, it's only for testing rbd).

Two new commits has been pushed, see https://github.com/kubernetes/kubernetes/pull/51608#issuecomment-326054083.

I will do more tests this weekend.

rootfs commented:
@cofyc can you create a `attacher.go` file and move the attach/detach implementation there? It makes review much easier. Thanks!

cofyc commented:
@rootfs 
OK. I amended commit directly, see new commint https://github.com/kubernetes/kubernetes/commit/c6bb49380259ab29bf124bf2139b306f45aefa38.

cofyc commented:
/test pull-kubernetes-unit

rootfs commented on pkg/volume/rbd/attacher.go:
```diff
@@ -0,0 +1,211 @@
+		return err
+	}
+	if cnt > 1 {
+		return fmt.Errorf("rbd: more than 1 reference counts at %s", deviceMountPath)
```
why error here?

cofyc replied:
In normal cases, `Detacher.UnmountDevice` [only be called once all bind mounts have been unmounted](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/volume.go#L201), `cnt` should always be 1 here. But if there is a bug somewhere or someone mount device on the node accidentally after [the check](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_generator.go#L579), `cnt` can be greater than 1, this is used to report this unexpected case.

Although kubelet will [let the caller to retry if device is still in used](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_generator.go#L591) even if `Detacher.UnmountDevice` returns nil, I think reporting this error is better than not (at least it's not wrong) and makes this function idempotent, only returns nil if everything is going well.

rootfs commented on pkg/volume/rbd/attacher.go:
```diff
@@ -0,0 +1,211 @@
+	if err != nil {
+		return err
+	}
+	if cnt > 1 {
```
call `UnmountPath` to umount it

cofyc replied:
hi, it's because `rbdDetacher.UnmountDevice` cannot use it to unmount device mount point, because `util.UnmountPath` will [delete the remaining directory if successful](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/util.go#L80). `rbdDetacher.UnmountDevice` needs to load RBD info from underlying directory later and remove lock if found. Almost the same logic as in [iscsi](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/iscsi/iscsi_util.go#L351).

cofyc commented:
/test pull-kubernetes-e2e-gce-etcd3

rootfs commented:
/approve

rootfs commented:
@cofyc see https://github.com/kubernetes/kubernetes/pull/51628

cofyc commented:
hi, @rootfs 

RBD plugin does not have this issue. In [AttachDisk](https://github.com/kubernetes/kubernetes/pull/51608/files#diff-36f18e327f36d95eb1333c4a18781184R274), I use [volutil.PathExists](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/util.go#L133) to check `globalPDPath`, will not return error if path does not exist (create it instead).

BTW, Why does iSCSI plugin mount the device at `globalPDPath` in [AttachDisk](https://github.com/kubernetes/kubernetes/blob/855c5e0e93421de09ca8c8dbfdf1fea01ae04737/pkg/volume/iscsi/iscsi_util.go#L331) (which is called by [WaitForAttach](https://github.com/kubernetes/kubernetes/blob/855c5e0e93421de09ca8c8dbfdf1fea01ae04737/pkg/volume/iscsi/attacher.go#L67)), and will try to mount again in [MountDevice](https://github.com/kubernetes/kubernetes/blob/855c5e0e93421de09ca8c8dbfdf1fea01ae04737/pkg/volume/iscsi/attacher.go#L90)?

As I think, Attacher/Detacher interfaces have `Attach` and `WaitForAttach` to attach device onto the machine, `MountDevice` to mount device at a point for later use (bind mounting to containers, etc), and`Detach` and `UnmountDevice` to do the opposite. There is no need to mount device at `globalPDPath` in `WaitForAttach` after implementing Attacher/Detacher interfaces.

rootfs commented:
/assign @mtanino 
/assign

cofyc commented:
@jingxu97 hi, Could you take a look at changes under [pkg/util/mount](https://github.com/kubernetes/kubernetes/pull/51608/commits/423a978eee7e37c7c28f0f06d23f70a9e7936dc9)? This fixs FakeMounter.IsLikelyNotMountPoint to return ErrNotExist if the
directory does not exist. Mounter.IsLikelyNotMountPoint interface
requires this, and RBD plugin depends on it.

cofyc commented:
hi, @deads2k , @lavalamp, @mikedanese 
Could you please approve changes under [cmd/kube-controller-manager](https://github.com/kubernetes/kubernetes/pull/51608/files#diff-57c787e8d6fc863735fea63e911331d8)? 
There is only one line change which adds RBD plugin into AttachableVolumePlugins list.

deads2k commented:
> hi, @deads2k , @lavalamp, @mikedanese
> Could you please approve changes under cmd/kube-controller-manager?
> There is only one line change which adds RBD plugin into AttachableVolumePlugins list.

We can't selectively approve packages, so we'll over approve.

Approve is intended to be done after the detailed review is over and the pull is lgtm'd.  Get through that and tag us again?

rootfs commented on pkg/volume/rbd/attacher.go:
```diff
@@ -0,0 +1,211 @@
+	mountOptions := volume.MountOptionFromSpec(spec, options...)
+	err = attacher.plugin.mounter.FormatAndMount(devicePath, deviceMountPath, volumeSource.FSType, mountOptions)
+	if err != nil {
+		return fmt.Errorf("rbd: failed to mount device %s at %s (fstype: %s), error %v", devicePath, deviceMountPath, volumeSource.FSType, err)
```
remove deviceMountPath before return

rootfs commented on pkg/volume/rbd/attacher.go:
```diff
@@ -0,0 +1,211 @@
+	if err != nil {
+		return fmt.Errorf("rbd: failed to mount device %s at %s (fstype: %s), error %v", devicePath, deviceMountPath, volumeSource.FSType, err)
+	}
+	glog.V(3).Infof("rbd: successfully mount device %s at %s (fstype: %s)", devicePath, deviceMountPath, volumeSource.FSType)
```
V(4)

rootfs commented:
@cofyc sorry for being late, thank you for being patient. Let's try to make this into 1.9.

This PR is getter better. I am asking one more thing.

Currently rbd checks if there is lock on the image and exits if so. With your attach/detach refactoring, we don't need to lock the image any more. We can remove all the fencing/defencing code and don't need to persist rbd.json as well.

To protect mapping images that are still being used by with nodes running old kubelet, we still need to check if there is any watchers on the image. AttachDisk only error if there is a confirmed watcher.

What do you think?

Commit:
RBD Plugin: Remove deviceMountPath before return on error
Attach.MountDevice.

Commit:
RBD Plugin: No need to acquire advisory lock any more!
With central attachdetach controller, we don't need to lock the image
any more. But for backward compatibility, we should:

1) Check if the image is still used by nodes running old kubelet in
attaching.
2) Clean old rbd.json file and remove lock if found in detaching.

cofyc commented:
hi, @rootfs

I think it is reasonable. In addition to check if the image is still used by nodes running old kubelet in
attaching, old rbd.json file and previous lock acquired by old kubelets should be cleaned also. Please review my [changes](https://github.com/kubernetes/kubernetes/pull/51608/commits/fe493649fe0a6b4cde5341c14641fcd51cb66e8d).

cofyc commented:
hi, @rootfs 
During development, I found `RBDUtil.rbdStatus` never return error even if command not found or `rbd status` never succeed. I think it should only check output of successful `rbd status` run to decide if there is a watcher on image. I submit [another PR](https://github.com/kubernetes/kubernetes/pull/54454) to address this issue.

rootfs commented:
@cofyc thanks 

I verified locally on my setup, the attach/detach code works, no lock held.

rootfs commented:
/lgtm

rootfs commented:
@deads2k @jingxu97 can you approve? thanks!

rootfs commented:
```release-note
RBD volume supports attach/detach 
```

cofyc commented:
@rootfs 
I rebased lots of change from https://github.com/kubernetes/kubernetes/pull/54302. I've done some tests again. If possible, could you verify it again on your setup before re-adding `lgtm` label？

rootfs commented:
/test pull-kubernetes-verify

rootfs commented:
ping @deads2k @jingxu97 
can you approve

jingxu97 commented:
/approve

rootfs commented:
/lgtm

brendandburns commented:
/approve
